docs/testing/snapshot-testing.md

```diff
@@ -0,0 +1,129 @@
+
+## Broken snapshots
+
+When a snapshot test fails, it just means that a component's rendering has changed. If that was unintended,  then the snapshot test just prevented a bug 😊
+
+However, if the change was intentional, follow these steps to update the snapshot:
+
+1. Run the following to update the snapshots:
+   ```sh
+   # --testPathPattern is optional but will be much faster by only running matching tests
+   npm run test-client -- --updateSnapshot --testPathPattern client/components
+   ```
+1. Review the diff and ensure the changes are expecte and intentional
+1. Commit
+
+## What are snapshots?
+
+Snapshots are just a representation of some data structure generated in our tests. Snapshots are stored in files and committed alongside the tests. When the tests are run, the data structure generated is compared with the snapshot on file.
+
+It's very easy to make a snapshot:
+
+```js
+test( 'foobar test', () => {
+  const foobar = { foo: 'bar' };
+  expect( foobar ).toMatchSnapshot();
+} );
+```
+
+This is the produced snapshot:
+
+```js
+exports[`test foobar test 1`] = `
+  Object {
+    "foo": "bar",
+  }
+`;
+```
+
+You should never interact with the snapshot directly, they are generated by tests. However you can see that it's a representation the test data.
+
+## Advantages
+
+
+* Trivial and concise to write tests
+* Protect against unintentional changes
+* Simple to work with
+
+## Disadvantages
+
+* Snapshot tests alone lack expressivity
+* Only tell us when things change
+* Like many tests, it's hard to work with anything non-deterministic or highly dependent on external factors
+
+## Use cases
+
+Snapshot are mostly targeted at component testing. They make us conscious of changes to a component's structure which makes them _ideal_ for refactoring. If a snapshot is kept up to date over the course of a series of commits, the snapshot diffs record the evolution of a component's structure. Pretty cool 😎
+
+```js
+import React from 'react';
+import { shallow } from 'enzyme';
+import MyComponent from '../my-component';
+import LocaleSuggestions from '../locale-suggestions';
+
+describe( 'MyComponent', () => {
+  test( 'should render', () => {
+    const wrapper = shallow( <MyComponent /> );
+    expect( wrapper ).toMatchSnapshot();
+  } );
+
+  test( 'should render with locale suggestions if locale is provided', () => {
+    const wrapper = shallow( <MyComponent locale="es" /> );
+    expect( wrapper ).toMatchSnapshot();
+    expect( wrapper.find( LocaleSuggestions ) ).toHaveLength( 1 );
+  } );
+} );
+```
+
+But the first snapshot tests introduced in Calypso were actually <a href="https://github.com/Automattic/wp-calypso/blob/e34d15f44c261fd7daa2212017e995883866d603/client/state/comments/test/selectors.js#L133-L142">reducer tests</a>. Brilliant! When we work with reducers we're often working with large data structures that we don't want to change unexpectedly.
```

Comment:
Why not [reducer tests](https://github.com/Automattic/wp-calypso/blob/e34d15f44c261fd7daa2212017e995883866d603/client/state/comments/test/selectors.js#L133-L142)? Asking out of my own curiosity.

Reply:
🤔  Apparently this doesn't read well, this is supposed to say:

> Reducer tests are also great fit for snapshots.

I'll adjust for clarity.

Reply:
My question was about the syntax used 😛 

On the second thought, your comment makes sense. Better to give a context for external contributors.

Reply:
I went with

> Reducer tests are also be a great fit for snapshots. The firsts snapshots introduced in Calypso were actually reducer tests. Reducers can be large, complex data structures that we don't want to change unexpectedly, exactly what snapshots excel at!